### PR TITLE
[search] Added centers table usage.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov @bykoianko PTAL

### CODE REVIEW

File: search/pre_ranker.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #include "search/pre_ranker.hpp"
 
+#include "search/lazy_centers_table.hpp"
```
Comment (bykoianko) [inline search/pre_ranker.cpp]: Sort please.

Comment (ygorshenin) [inline search/pre_ranker.cpp]: Done

File: search/mwm_context.hpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "search/house_to_street_table.hpp"
+#include "search/lazy_centers_table.hpp"
```
Comment (bykoianko) [inline search/mwm_context.hpp]: #include "std/shared_ptr.hpp"
#include "std/string.hpp"
?

File: generator/centers_table_builder.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  {
+    auto const & platform = GetPlatform();
+    FilesContainerR rcont(platform.GetReader(filename, "f"));
+    if (!forceRebuild && rcont.IsExist(CENTERS_FILE_TAG))
+      return true;
+
+    feature::DataHeader header(rcont);
+
+    version::MwmTraits traits(header.GetFormat());
```
Comment (bykoianko) [inline generator/centers_table_builder.cpp]: const?

File: generator/centers_table_builder.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    if (!traits.HasOffsetsTable())
+    {
+      LOG(LERROR, (filename, "does not have an offsets table!"));
+      return false;
+    }
+    auto table = feature::FeaturesOffsetsTable::Load(rcont);
```
Comment (bykoianko) [inline generator/centers_table_builder.cpp]: const?

File: generator/centers_table_builder.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    if (!table)
+    {
+      LOG(LERROR, ("Can't load offsets table from:", filename));
+      return false;
+    }
+
+    FeaturesVector features(rcont, header, table.get());
```
Comment (bykoianko) [inline generator/centers_table_builder.cpp]: const?

File: search/lazy_centers_table.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+{
+public:
+  enum State
+  {
+    STATE_NOT_LOADED,
+    STATE_LOADED,
+    STATE_FAILED
+  };
+
+  LazyCentersTable(MwmValue & value);
```
Comment (bykoianko) [inline search/lazy_centers_table.hpp]: explicit?

File: search/mwm_context.hpp (hunk 2 above)
Comment (ygorshenin) [inline search/mwm_context.hpp]: Done

File: generator/centers_table_builder.cpp (hunk 3 above)
Comment (ygorshenin) [inline generator/centers_table_builder.cpp]: Done

File: generator/centers_table_builder.cpp (hunk 4 above)
Comment (ygorshenin) [inline generator/centers_table_builder.cpp]: Done

File: generator/centers_table_builder.cpp (hunk 5 above)
Comment (ygorshenin) [inline generator/centers_table_builder.cpp]: Done

File: search/lazy_centers_table.hpp (hunk 6 above)
Comment (ygorshenin) [inline search/lazy_centers_table.hpp]: Done

**Comment by ygorshenin** [comment]
PTAL

File: generator/generator_tool/generator_tool.cpp [hunk 7]
```diff
@@ -127,8 +129,8 @@ int main(int argc, char ** argv)
 
   // Load classificator only when necessary.
   if (FLAGS_make_coasts || FLAGS_generate_features || FLAGS_generate_geometry ||
-      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_calc_statistics ||
-      FLAGS_type_statistics || FLAGS_dump_types || FLAGS_dump_prefixes ||
+      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_generate_centers_table ||
```
Comment (mpimenov) [inline generator/generator_tool/generator_tool.cpp]: Why do you need classificator here?

File: indexer/centers_table.cpp [hunk 8]
```diff
@@ -262,17 +263,21 @@ unique_ptr<CentersTable> CentersTable::Load(Reader & reader,
...
 {
+  if (!m_ids.empty())
+    CHECK_LESS(m_ids.back(), featureId, ());
+
+  m_centers.push_back(PointD2PointU(center, m_codingParams.GetCoordBits()));
+  m_ids.push_back(featureId);
 }
 
-CentersTableBuilder::~CentersTableBuilder()
+void CentersTableBuilder::Freeze(Writer & writer)
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: It seems to me it's possible to make this method constant. Is it so?

File: generator/centers_table_builder.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace indexer
+{
+bool BuildCentersTableFromDataFile(string const & filename, bool forceRebuild)
+{
+  search::CentersTableBuilder builder;
+
+  {
+    auto const & platform = GetPlatform();
+    FilesContainerR rcont(platform.GetReader(filename, "f"));
```
Comment (bykoianko) [inline generator/centers_table_builder.cpp]: Platform::GetReader(...) calls Platform::ReadPathForFile(...) and Platform::ReadPathForFile could throw MYTHROW(FileAbsentException, ... .
Could you please write try/catch in BuildCentersTableFromDataFile?

File: generator/generator_tool/generator_tool.cpp (hunk 7 above)
Comment (ygorshenin) [inline generator/generator_tool/generator_tool.cpp]: mpimenov <notifications@github.com> writes:

> In generator/generator_tool/generator_tool.cpp:
>
>> @@ -127,8 +129,8 @@ int main(int argc, char ** argv)
>>  
>>    // Load classificator only when necessary.
>>    if (FLAGS_make_coasts || FLAGS_generate_features || FLAGS_generate_geometry ||
>> -      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_calc_statistics ||
>> -      FLAGS_type_statistics || FLAGS_dump_types || FLAGS_dump_prefixes ||
>> +      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_generate_centers_table ||
>
> Why do you need classificator here?

Sorry, I didn't get your question. As you can see, I've just added a new
flag check to the condition.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

File: indexer/centers_table.cpp (hunk 8 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: No problem, done.

-- 
Best regards, Yuri

File: generator/generator_tool/generator_tool.cpp (hunk 7 above)
Comment (mpimenov) [inline generator/generator_tool/generator_tool.cpp]: This condition guards loading of classificator. Why did you add the new flag check?

File: generator/centers_table_builder.cpp (hunk 9 above)
Comment (ygorshenin) [inline generator/centers_table_builder.cpp]: Vladimir Byko-Ianko <notifications@github.com> writes:

> In generator/centers_table_builder.cpp:
>
>> +#include "coding/file_container.hpp"
>> +
>> +#include "platform/mwm_traits.hpp"
>> +#include "platform/platform.hpp"
>> +
>> +#include "defines.hpp"
>> +
>> +namespace indexer
>> +{
>> +bool BuildCentersTableFromDataFile(string const & filename, bool forceRebuild)
>> +{
>> +  search::CentersTableBuilder builder;
>> +
>> +  {
>> +    auto const & platform = GetPlatform();
>> +    FilesContainerR rcont(platform.GetReader(filename, "f"));
>
> Platform::GetReader(...) calls Platform::ReadPathForFile(...) and Platform::ReadPathForFile could throw MYTHROW(FileAbsentException, ... .
> Could you please write try/catch in BuildCentersTableFromDataFile?

Usage of platform is replaced by direct creation of FilesContainerR from
filename - it's possible. But this just hides the problem, not fixes it.
Note that Readers and Writers may throw exceptions too (no space left on
device, corrupted filesystem, etc), but this code is going to be used on
server, during maps generation, so do we need to catch and properly
handle exceptions here? Just for consistency with the method signature
(returns false in case of errors), I'm going to add an exception handler
here, but the method still won't be noexcept.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

**Comment by bykoianko** [comment]
Please rebase the branch.

**Comment by bykoianko** [comment]
DUDE JTTP

File: generator/generator_tool/generator_tool.cpp (hunk 7 above)
Comment (ygorshenin) [inline generator/generator_tool/generator_tool.cpp]: mpimenov <notifications@github.com> writes:

> In generator/generator_tool/generator_tool.cpp:
>
>> @@ -127,8 +129,8 @@ int main(int argc, char ** argv)
>>  
>>    // Load classificator only when necessary.
>>    if (FLAGS_make_coasts || FLAGS_generate_features || FLAGS_generate_geometry ||
>> -      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_calc_statistics ||
>> -      FLAGS_type_statistics || FLAGS_dump_types || FLAGS_dump_prefixes ||
>> +      FLAGS_generate_index || FLAGS_generate_search_index || FLAGS_generate_centers_table ||
>
> This condition guards loading of classificator. Why did you add the new flag check?
>
Because all stuff that deals with features requires classifier. In any
case, the flag is removed.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

**Comment by ygorshenin** [comment]
Added a test. PTAL @bykoianko @mpimenov

File: search/search_integration_tests/pre_ranker_test.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+using namespace generator::tests_support;
+using namespace search::tests_support;
+
+namespace search
+{
+namespace
+{
+// Returns the boundary such that at least |size| elements from
+// |distance| are less or equal than the boundary.
```
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: are less than or equal to the boundary

File: search/search_integration_tests/pre_ranker_test.cpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+double GetBoundary(vector<double> const & distances, size_t size)
+{
+  my::limited_priority_queue<double> queue(size);
+  for (auto const & distance : distances)
+    queue.push(distance);
+  return queue.empty() ? 0.0 : queue.top();
+}
+
+class FakeRanker : public Ranker
```
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: s/FakeRanker/TestRanker/ ?

File: search/search_integration_tests/pre_ranker_test.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+using namespace generator::tests_support;
+using namespace search::tests_support;
+
+namespace search
+{
+namespace
+{
+// Returns the boundary such that at least |size| elements from
+// |distance| are less or equal than the boundary.
+double GetBoundary(vector<double> const & distances, size_t size)
```
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: ```assert(size <= distances.size())``` ?

File: search/search_integration_tests/processor_test.cpp [hunk 13]
```diff
@@ -492,7 +492,6 @@ UNIT_CLASS_TEST(ProcessorTest, TestPostcodes)
   // Tests that postcode is added to the search index.
   {
     MwmContext context(m_engine.GetMwmHandleById(countryId));
-    TEST(context.IsAlive(), ());
```
Comment (mpimenov) [inline search/search_integration_tests/processor_test.cpp]: А что поменялось?

**Comment by bykoianko** [comment]
LGTM

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 12 above)
Comment (ygorshenin) [inline search/search_integration_tests/pre_ranker_test.cpp]: mpimenov <notifications@github.com> writes:

> In search/search_integration_tests/pre_ranker_test.cpp:
>
>> +#include "base/stl_add.hpp"
>> +
>> +#include "std/algorithm.hpp"
>> +#include "std/iterator.hpp"
>> +#include "std/vector.hpp"
>> +
>> +using namespace generator::tests_support;
>> +using namespace search::tests_support;
>> +
>> +namespace search
>> +{
>> +namespace
>> +{
>> +// Returns the boundary such that at least |size| elements from
>> +// |distance| are less or equal than the boundary.
>> +double GetBoundary(vector<double> const & distances, size_t size)
>
> assert(size <= distances.size()) ?

Why? Does it change the specification?

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

File: search/search_integration_tests/processor_test.cpp (hunk 13 above)
Comment (ygorshenin) [inline search/search_integration_tests/processor_test.cpp]: mpimenov <notifications@github.com> writes:

> In search/search_integration_tests/processor_test.cpp:
>
>> @@ -492,7 +492,6 @@ UNIT_CLASS_TEST(ProcessorTest, TestPostcodes)
>>    // Tests that postcode is added to the search index.
>>    {
>>      MwmContext context(m_engine.GetMwmHandleById(countryId));
>> -    TEST(context.IsAlive(), ());
>
> А что поменялось?

This CL removes MwmContext::IsAlive(). MwmContext::IsAlive() is broken
by design.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 11 above)
Comment (ygorshenin) [inline search/search_integration_tests/pre_ranker_test.cpp]: mpimenov <notifications@github.com> writes:

> In search/search_integration_tests/pre_ranker_test.cpp:
>
>> +
>> +namespace search
>> +{
>> +namespace
>> +{
>> +// Returns the boundary such that at least |size| elements from
>> +// |distance| are less or equal than the boundary.
>> +double GetBoundary(vector<double> const & distances, size_t size)
>> +{
>> +  my::limited_priority_queue<double> queue(size);
>> +  for (auto const & distance : distances)
>> +    queue.push(distance);
>> +  return queue.empty() ? 0.0 : queue.top();
>> +}
>> +
>> +class FakeRanker : public Ranker
>
> s/FakeRanker/TestRanker/ ?
Real difference?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

**Comment by ygorshenin** [comment]
PTAL

File: search/search_integration_tests/processor_test.cpp (hunk 13 above)
Comment (mpimenov) [inline search/search_integration_tests/processor_test.cpp]: Ок.

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 11 above)
Comment (mpimenov) [inline search/search_integration_tests/pre_ranker_test.cpp]: Well, I did not insist.

Comment (ygorshenin) [inline search/search_integration_tests/pre_ranker_test.cpp]: mpimenov <notifications@github.com> writes:

> In search/search_integration_tests/pre_ranker_test.cpp:
>
>> +
>> +namespace search
>> +{
>> +namespace
>> +{
>> +// Returns the boundary such that at least |size| elements from
>> +// |distance| are less or equal than the boundary.
>> +double GetBoundary(vector<double> const & distances, size_t size)
>> +{
>> +  my::limited_priority_queue<double> queue(size);
>> +  for (auto const & distance : distances)
>> +    queue.push(distance);
>> +  return queue.empty() ? 0.0 : queue.top();
>> +}
>> +
>> +class FakeRanker : public Ranker
>
> Well, I did not insist.
Too late...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

**Comment by ygorshenin** [comment]
PTAL

File: search/search_integration_tests/pre_ranker_test.cpp (hunk 12 above)
Comment (ygorshenin) [inline search/search_integration_tests/pre_ranker_test.cpp]: mpimenov <notifications@github.com> writes:

> In search/search_integration_tests/pre_ranker_test.cpp:
>
>> +#include "base/stl_add.hpp"
>> +
>> +#include "std/algorithm.hpp"
>> +#include "std/iterator.hpp"
>> +#include "std/vector.hpp"
>> +
>> +using namespace generator::tests_support;
>> +using namespace search::tests_support;
>> +
>> +namespace search
>> +{
>> +namespace
>> +{
>> +// Returns the boundary such that at least |size| elements from
>> +// |distance| are less or equal than the boundary.
>> +double GetBoundary(vector<double> const & distances, size_t size)
>
> assert(size <= distances.size()) ?

Okay, this is a strong signal that method does something strange if it's
hard to invent a clear comment to it. Added CHECK().

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri

**Comment by mpimenov** [comment]
LGTM
